src/vm/ssa.nit

```diff
@@ -0,0 +1,1140 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Compute Single-Static Assignment form from an AST
+module ssa
+
+import variables_numbering
+import astbuilder
+
+redef class VirtualMachine
+
+	# The currently analyzed APropdef
+	var current_propdef: APropdef
+
+	redef fun new_frame(node, mpropdef, args)
+	do
+		# Save the current propdef
+		if node isa APropdef then current_propdef = node
+
+		return super
+	end
+end
+
+# Represent a sequence of the program
+# A basic block is composed of several instruction without a jump
+class BasicBlock
+	# First instruction of the basic block
+	var first: ANode is noinit
+
+	# Last instruction of the basic block
+	var last: ANode is noinit
+
+	# Direct successors
+	var successors = new Array[BasicBlock]
+
+	# Direct predecessors
+	var predecessors = new Array[BasicBlock]
+
+	# Parts of AST that contain a read to a variable
+	var read_sites = new Array[AVarFormExpr]
+
+	# Parts of AST that contain a write to a variable
+	var write_sites = new Array[AVarFormExpr]
+
+	# Parts of AST that contain a variable access (read or write)
+	var variables_sites = new Array[AExpr]
+
+	# The iterated dominance frontier of this block
+	# i.e. the set of blocks this block dominate directly or indirectly
+	var dominance_frontier: Array[BasicBlock] = new Array[BasicBlock] is lazy
+
+	# Self is the old block to link to the new,
+	# The two blocks are not linked if the current ends with a `AReturnExpr` or `ABreakExpr`
+	# i.e. self is the predecessor of `successor`
+	# `successor` The successor block
+	fun link(successor: BasicBlock)
+	do
+		# Do not link the two blocks if the current block end with a return
+		if last isa AReturnExpr then return
+
+		if last isa ABreakExpr then return
+
+		successors.add(successor)
+		successor.predecessors.add(self)
+	end
+
+	# Self is the old block to link to the new
+	# i.e. self is the predecessor of `successor`
+	# `successor` The successor block
+	fun link_special(successor: BasicBlock)
+	do
+		# Link the two blocks even if the current block ends with a return or a break
+		successors.add(successor)
+		successor.predecessors.add(self)
+	end
+
+	# Add the `block` to the dominance frontier of this block
+	fun add_df(block: BasicBlock)
+	do
+		# Add this block recursively in super-blocks to compute the iterated
+		# dominance frontier
+		dominance_frontier.add(block)
+
+		for successor in block.successors do
+			# If this successor has not already been add to the dominance frontier
+			if not dominance_frontier.has(successor) then
+				add_df(successor)
+			end
+		end
+	end
+
+	# Compute recursively the dominance frontier of self block and its successors
+	private fun compute_df
+	do
+		# Treat each block only one time
+		df_computed = true
+
+		for s in successors do
+			add_df(s)
+
+			if not s.df_computed then s.compute_df
+		end
+	end
+
+	# Used to dump the BasicBlock to dot
+	var treated: Bool = false
```

Comment:
My mistake, I need to add a second variable and not used this one. Actually I do a lot of things recursively and I often need some variables to treat loops in recursions